File: src/amrte/web/dataset_replay.py
"""Read-only Dataset & Replay integrity projection for the AMRTE web console."""

from __future__ import annotations

from typing import Any

from amrte.core.constants import AMRTE_VERSION

VERSION = AMRTE_VERSION


DATASET_SERVICE_NAMES = (
    "market_data",
    "market_data_provider",
    "dataset",
    "dataset_service",
)

EVENT_SERVICE_NAMES = (
    "events",
    "event_provider",
    "historical_events",
)

REPLAY_SERVICE_NAMES = (
    "replay",
    "replay_service",
    "backtest",
    "simulation",
)


IMPLEMENTED_CAPABILITIES = (
    {
        "capability": "DATASET_IDENTITY",
        "component": "DeterministicMarketDataProvider",
        "implemented": True,
        "runtime_authoritative": False,
    },
    {
        "capability": "DATASET_FINGERPRINT",
        "component": "DeterministicMarketDataProvider",
        "implemented": True,
        "runtime_authoritative": False,
    },
    {
        "capability": "AS_OF_MARKET_DATA",
        "component": "DeterministicMarketDataProvider",
        "implemented": True,
        "runtime_authoritative": False,
    },
    {
        "capability": "DATASET_VALIDATION_ADMISSION",
        "component": "MarketDataService",
        "implemented": True,
        "runtime_authoritative": False,
    },
    {
        "capability": "HISTORICAL_EVENT_AS_OF",
        "component": "DeterministicEventProvider",
        "implemented": True,
        "runtime_authoritative": False,
    },
    {
        "capability": "TEMPORAL_NO_LOOK_AHEAD",
        "component": "MasterConfiguration",
        "implemented": True,
        "runtime_authoritative": True,
    },
)


def _enum_name(value: Any) -> Any:
    if hasattr(value, "name"):
        return value.name
    return value


def _configuration_values(runtime: Any) -> dict[str, Any]:
    engine = runtime.engine

    if engine is None:
        return {}

    provider = engine.registry.require(
        "configuration"
    )

    loaded = provider.load()

    if not loaded.success:
        return {}

    snapshot = loaded.value

    return {
        key: _enum_name(value)
        for key, value in snapshot.values.items()
    }


def _registered_services(
    runtime: Any,
) -> dict[str, Any]:
    engine = runtime.engine

    if engine is None:
        return {}

    result = {}

    for name in engine.registry.names():
        service = engine.registry.require(name)

        result[name] = {
            "type": type(service).__name__,
            "module": type(service).__module__,
        }

    return result


def _first_registered(
    services: dict[str, Any],
    names: tuple[str, ...],
) -> str | None:
    for name in names:
        if name in services:
            return name

    return None


def build_dataset_replay_integrity(
    runtime: Any,
) -> dict[str, Any]:
    """Project existing runtime/configuration state without mutation."""

    values = _configuration_values(runtime)
    services = _registered_services(runtime)

    dataset_service = _first_registered(
        services,
        DATASET_SERVICE_NAMES,
    )

    event_service = _first_registered(
        services,
        EVENT_SERVICE_NAMES,
    )

    replay_service = _first_registered(
        services,
        REPLAY_SERVICE_NAMES,
    )

    checkpoint = runtime.last_checkpoint

    persistence_dataset_id = (
        checkpoint.dataset_id
        if checkpoint is not None
        else None
    )

    persistence_fingerprint = (
        checkpoint.dataset_fingerprint
        if checkpoint is not None
        else None
    )

    configured_dataset_id = values.get(
        "backtest.dataset_id"
    )

    market_provider_type = values.get(
        "market_data.provider_type"
    )

    event_provider_type = values.get(
        "events.provider_type"
    )

    no_look_ahead = values.get(
        "backtest.no_look_ahead"
    )

    runtime_mode = values.get(
        "runtime.mode"
    )

    execution = services.get(
        "execution",
        {},
    )

    execution_provider = execution.get(
        "type"
    )

    return {
        "version": VERSION,

        "mode": "READ_ONLY",

        "research_only": True,

        "runtime": {
            "state": runtime.state,
            "environment": (
                runtime.engine.environment.name
                if runtime.engine is not None
                else "UNKNOWN"
            ),
            "recovery_epoch": runtime.recovery_epoch,
        },

        "dataset_identity": {
            "configured_dataset_id":
                configured_dataset_id,

            "persistence_dataset_id":
                persistence_dataset_id,

            "persistence_dataset_fingerprint":
                persistence_fingerprint,

            "runtime_dataset_provider_registered":
                dataset_service is not None,

            "runtime_dataset_service":
                dataset_service,
        },

        "provider_policy": {
            "market_data_provider_type":
                market_provider_type,

            "event_provider_type":
                event_provider_type,

            "market_data_offline":
                market_provider_type
                == "DETERMINISTIC_OFFLINE",

            "events_offline":
                event_provider_type
                == "DETERMINISTIC_OFFLINE",

            "event_provider_registered":
                event_service is not None,

            "runtime_event_service":
                event_service,
        },

        "temporal_integrity": {
            "no_look_ahead":
                no_look_ahead is True,

            "configured_value":
                no_look_ahead,

            "as_of_market_data_supported":
                True,

            "as_of_historical_events_supported":
                True,
        },

        "replay_authority": {
            "authoritative_replay_service_registered":
                replay_service is not None,

            "runtime_replay_service":
                replay_service,

            "replay_control_available":
                False,

            "mutation_enabled":
                False,

            "reason": (
                "AUTHORITATIVE_REPLAY_SERVICE_NOT_REGISTERED"
                if replay_service is None
                else "WEB_REPLAY_MUTATION_NOT_ENABLED"
            ),
        },

        "implemented_capabilities": [
            {
                **item,
                "runtime_authoritative": (
                    item["runtime_authoritative"]
                    or (
                        item["component"]
                        == "DeterministicMarketDataProvider"
                        and dataset_service is not None
                    )
                    or (
                        item["component"]
                        == "MarketDataService"
                        and dataset_service is not None
                    )
                    or (
                        item["component"]
                        == "DeterministicEventProvider"
                        and event_service is not None
                    )
                ),
            }
            for item in IMPLEMENTED_CAPABILITIES
        ],

        "execution_boundary": {
            "capability": "PROHIBITED",

            "provider":
                execution_provider,

            "financial_execution_available":
                False,
        },

        "configuration": {
            "runtime_mode":
                runtime_mode,

            "configuration_snapshot_id": (
                checkpoint.configuration_snapshot_id
                if checkpoint is not None
                else None
            ),

            "configuration_hash": (
                checkpoint.configuration_hash
                if checkpoint is not None
                else None
            ),
        },

        "registered_services": services,
    }

File: src/amrte/web/static/dataset-replay.js
"use strict";

function text(value) {
    if (
        value === null ||
        value === undefined ||
        value === ""
    ) {
        return "?";
    }

    return String(value);
}

function state(value) {
    return value ? "YES" : "NO";
}

function registered(value) {
    return value
        ? "REGISTERED"
        : "NOT REGISTERED";
}

function available(value) {
    return value
        ? "AVAILABLE"
        : "UNAVAILABLE";
}

function detailRow(label, value) {
    return `
        <div class="detail-row">
            <span>${label}</span>
            <strong>${text(value)}</strong>
        </div>
    `;
}

function renderDataset(data) {
    const dataset = data.dataset_identity;

    document.getElementById(
        "dataset-details"
    ).innerHTML = [
        detailRow(
            "Configured dataset",
            dataset.configured_dataset_id
        ),
        detailRow(
            "Persistence dataset",
            dataset.persistence_dataset_id
        ),
        detailRow(
            "Persistence fingerprint",
            dataset.persistence_dataset_fingerprint
        ),
        detailRow(
            "Runtime provider",
            registered(
                dataset.runtime_dataset_provider_registered
            )
        ),
        detailRow(
            "Runtime dataset service",
            dataset.runtime_dataset_service
        ),
    ].join("");
}

function renderProviderPolicy(data) {
    const policy = data.provider_policy;

    document.getElementById(
        "provider-details"
    ).innerHTML = [
        detailRow(
            "Market-data provider",
            policy.market_data_provider_type
        ),
        detailRow(
            "Market data offline",
            state(policy.market_data_offline)
        ),
        detailRow(
            "Event provider",
            policy.event_provider_type
        ),
        detailRow(
            "Events offline",
            state(policy.events_offline)
        ),
        detailRow(
            "Runtime event provider",
            registered(
                policy.event_provider_registered
            )
        ),
        detailRow(
            "Runtime event service",
            policy.runtime_event_service
        ),
    ].join("");
}

function renderTemporal(data) {
    const temporal = data.temporal_integrity;

    document.getElementById(
        "temporal-details"
    ).innerHTML = [
        detailRow(
            "No look ahead",
            state(temporal.no_look_ahead)
        ),
        detailRow(
            "Configured value",
            state(temporal.configured_value)
        ),
        detailRow(
            "As-of market data",
            available(
                temporal.as_of_market_data_supported
            )
        ),
        detailRow(
            "As-of historical events",
            available(
                temporal
                    .as_of_historical_events_supported
            )
        ),
    ].join("");
}

function renderReplay(data) {
    const replay = data.replay_authority;

    document.getElementById(
        "replay-details"
    ).innerHTML = [
        detailRow(
            "Authoritative service",
            registered(
                replay
                    .authoritative_replay_service_registered
            )
        ),
        detailRow(
            "Runtime replay service",
            replay.runtime_replay_service
        ),
        detailRow(
            "Replay control",
            available(
                replay.replay_control_available
            )
        ),
        detailRow(
            "Mutation enabled",
            state(replay.mutation_enabled)
        ),
        detailRow(
            "Reason",
            replay.reason
        ),
    ].join("");
}

function renderConfiguration(data) {
    const config = data.configuration;

    document.getElementById(
        "configuration-details"
    ).innerHTML = [
        detailRow(
            "Runtime mode",
            config.runtime_mode
        ),
        detailRow(
            "Snapshot ID",
            config.configuration_snapshot_id
        ),
        detailRow(
            "Configuration hash",
            config.configuration_hash
        ),
    ].join("");
}

function renderExecution(data) {
    const boundary = data.execution_boundary;

    document.getElementById(
        "execution-details"
    ).innerHTML = [
        detailRow(
            "Capability",
            boundary.capability
        ),
        detailRow(
            "Provider",
            boundary.provider
        ),
        detailRow(
            "Financial execution",
            boundary.financial_execution_available
                ? "AVAILABLE"
                : "NOT AVAILABLE"
        ),
    ].join("");
}

function renderCapabilities(data) {
    const capabilities =
        data.implemented_capabilities;

    const container = document.getElementById(
        "implemented-capabilities"
    );

    document.getElementById(
        "capability-count"
    ).textContent =
        `${capabilities.length} capabilities`;

    container.innerHTML = capabilities.map(
        capability => `
            <article class="control-card">
                <div class="control-card-header">
                    <strong>
                        ${text(capability.capability)}
                    </strong>

                    <span class="status-pill ${
                        capability.runtime_authoritative
                            ? ""
                            : "unavailable"
                    }">
                        ${
                            capability.runtime_authoritative
                                ? "AUTHORITATIVE"
                                : "NON-AUTHORITATIVE"
                        }
                    </span>
                </div>

                <span class="control-category">
                    ${text(capability.component)}
                </span>

                <p>
                    Implementation:
                    ${
                        capability.implemented
                            ? "AVAILABLE"
                            : "UNAVAILABLE"
                    }
                </p>
            </article>
        `
    ).join("");
}

function renderRegisteredServices(data) {
    const services = data.registered_services;

    const container = document.getElementById(
        "registered-services"
    );

    const entries = Array.isArray(services)
        ? services.map(
            service => [
                service.name,
                service,
            ]
        )
        : Object.entries(services);

    document.getElementById(
        "service-count"
    ).textContent =
        `${entries.length} services`;

    container.innerHTML = entries.map(
        ([name, service]) => `
            <article class="service-registry-card">
                <div class="service-registry-header">
                    <div class="service-registry-icon">
                        ${text(name)
                            .slice(0, 1)
                            .toUpperCase()}
                    </div>

                    <div class="service-registry-title">
                        <strong>
                            ${text(name)}
                        </strong>

                        <span class="service-status">
                            REGISTERED
                        </span>
                    </div>
                </div>

                <div class="service-registry-detail">
                    <span>Implementation</span>
                    <strong>
                        ${text(service.type)}
                    </strong>
                </div>

                <div class="service-registry-detail">
                    <span>Module</span>
                    <code>
                        ${text(service.module)}
                    </code>
                </div>
            </article>
        `
    ).join("");
}

async function loadDatasetReplay() {
    const response = await fetch(
        "/api/v1/dataset-replay",
        {
            method: "GET",
            headers: {
                "Accept": "application/json",
            },
        }
    );

    if (!response.ok) {
        throw new Error(
            `Dataset & Replay API returned ${response.status}`
        );
    }

    const data = await response.json();

    document.getElementById(
        "runtime-state"
    ).textContent =
        data.runtime.state;

    const runtimeStatusDot =
        document.getElementById(
            "runtime-status-dot"
        );

    if (runtimeStatusDot) {
        runtimeStatusDot.classList.toggle(
            "running",
            data.runtime.state === "RUNNING"
        );
    }

    document.getElementById(
        "metric-dataset"
    ).textContent =
        text(
            data.dataset_identity
                .configured_dataset_id
        );

    document.getElementById(
        "metric-provider"
    ).textContent =
        text(
            data.provider_policy
                .market_data_provider_type
        );

    document.getElementById(
        "metric-temporal"
    ).textContent =
        data.temporal_integrity.no_look_ahead
            ? "ENFORCED"
            : "NOT ENFORCED";

    document.getElementById(
        "metric-replay"
    ).textContent =
        data.replay_authority
            .authoritative_replay_service_registered
            ? "REGISTERED"
            : "NOT REGISTERED";

    renderDataset(data);
    renderProviderPolicy(data);
    renderTemporal(data);
    renderReplay(data);
    renderConfiguration(data);
    renderExecution(data);
    renderCapabilities(data);
    renderRegisteredServices(data);
}

loadDatasetReplay().catch(error => {
    document.getElementById(
        "runtime-state"
    ).textContent = "ERROR";

    console.error(
        "AMRTE Dataset & Replay:",
        error
    );
});
